pilot/pkg/serviceregistry/kube/controller/controller.go

```diff
@@ -467,24 +463,48 @@ func (c *Controller) Stop() {
 
 // Services implements a service catalog operation
 func (c *Controller) Services() ([]*model.Service, error) {
-	var k8sNodes []*kubernetesNode
-	var extAddresses []string
-	rawNodes := c.filteredNodeInformer.GetStore().List()
-	for _, i := range rawNodes {
-		machine := i.(*v1.Node)
-		for _, address := range machine.Status.Addresses {
-			if address.Type == v1.NodeExternalIP && address.Address != "" {
-				k8sNodes = append(k8sNodes,
-					&kubernetesNode{address: address.Address, labels: machine.Labels})
-				extAddresses = append(extAddresses, address.Address)
-			}
-		}
-	}
-
 	c.Lock()
 	out := make([]*model.Service, 0, len(c.servicesMap))
 	for _, svc := range c.servicesMap {
-		if isNodePortService(svc) {
+		out = append(out, svc)
+	}
+	c.Unlock()
+	sort.Slice(out, func(i, j int) bool { return out[i].Hostname < out[j].Hostname })
+
+	return out, nil
+}
+
+// GetService implements a service catalog operation by hostname specified.
+func (c *Controller) GetService(hostname host.Name) (*model.Service, error) {
+	c.RLock()
+	svc := c.servicesMap[hostname]
+	c.RUnlock()
+	return svc, nil
+}
+
+// updateServiceExternalAddr updates ClusterExternalAddresses for ingress gateway service of nodePort type
+func (c *Controller) updateServiceExternalAddr(svcs ...*model.Service) {
+	// node event, update all nodePort gateway services
+	if len(svcs) == 0 {
+		svcs, _ = c.Services()
+	}
+	for _, svc := range svcs {
+		if isNodePortGatewayService(svc) {
+			// update external address
+			var k8sNodes []*kubernetesNode
+			var extAddresses []string
+			rawNodes := c.filteredNodeInformer.GetStore().List()
+			for _, i := range rawNodes {
+				machine := i.(*v1.Node)
+				for _, address := range machine.Status.Addresses {
+					if address.Type == v1.NodeExternalIP && address.Address != "" {
+						k8sNodes = append(k8sNodes,
+							&kubernetesNode{address: address.Address, labels: machine.Labels})
+						extAddresses = append(extAddresses, address.Address)
+					}
+				}
+			}
```

Comment:
This logic needs to be moved out. It’s one time computation for all services

Reply:
I donot know what i was thinking